drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,713 @@
+{
+	int ret;
+
+	ret = i2c_esp32_transmit(dev);
+	if (!ret) {
+		return i2c_esp32_wait(dev, wait_cmd);
+	}
+
+	return ret;
+}
+
+static volatile struct i2c_esp32_cmd *
+i2c_esp32_write_addr(struct device *dev,
+		     volatile struct i2c_esp32_cmd *cmd,
+		     struct i2c_msg *msg,
+		     u16_t addr)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t addr_len = 1;
+
+	i2c_esp32_reset_fifo(config);
+
+	sys_write32(addr & I2C_FIFO_RDATA, I2C_DATA_APB_REG(config->index));
+	if (data->dev_config & I2C_ADDR_10_BITS) {
+		sys_write32(I2C_DATA_APB_REG(config->index),
+			    (addr >> 8) & I2C_FIFO_RDATA);
+		addr_len++;
+	}
+
+	*cmd++ = (struct i2c_esp32_cmd) {
+		.opcode = I2C_ESP32_OP_WRITE,
+		.ack_en = true,
+		.num_bytes = addr_len,
+	};
+
+	msg->len += addr_len;
+
+	return cmd;
+}
+
+static int i2c_esp32_read_msg(struct device *dev, u16_t addr,
+			      struct i2c_msg msg)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	volatile struct i2c_esp32_cmd *cmd = (void *)I2C_COMD0_REG(config->index);
+	u32_t i;
+	int ret;
+
+	/* Set the R/W bit to R */
+	addr |= BIT(0);
+
+	if (msg.flags & I2C_MSG_RESTART) {
+		*cmd++ = (struct i2c_esp32_cmd) {
+			.opcode = I2C_ESP32_OP_RSTART
+		};
+	}
+
+	cmd = i2c_esp32_write_addr(dev, cmd, &msg, addr);
+
+	for (; msg.len; cmd = (void *)I2C_COMD0_REG(config->index)) {
+		volatile struct i2c_esp32_cmd *wait_cmd = NULL;
+		u32_t to_read = max(I2C_ESP32_BUFFER_SIZE, msg.len - 1);
+
+		/* Might be the last byte, in which case, `to_read` will
+		 * be 0 here.  See comment below.
+		 */
+		if (to_read) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = to_read,
+			};
+		}
+
+		/* I2C master won't acknowledge the last byte read from the
+		 * slave device.  Divide the read command in two segments as
+		 * recommended by the ESP32 Technical Reference Manual.
+		 */
+		if (msg.len - to_read <= 1) {
+			/* Read the last byte and explicitly ask for an
+			 * acknowledgement.
+			 */
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = 1,
+				.ack_val = true,
+			};
+
+			/* Account for the `msg.len - 1` when clamping
+			 * transmission length to FIFO buffer size.
+			 */
+			to_read++;
+
+			if (msg.flags & I2C_MSG_STOP) {
+				wait_cmd = cmd;
+				*cmd++ = (struct i2c_esp32_cmd) {
+					.opcode = I2C_ESP32_OP_STOP
+				};
+			}
+		}
+		if (!wait_cmd) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_END
+			};
+		}
+
+		ret = i2c_esp32_transmit_wait(dev, wait_cmd);
+		if (ret < 0) {
+			return ret;
+		}
+
+		for (i = 0; i < to_read; i++) {
+			u32_t v = sys_read32(I2C_DATA_APB_REG(config->index));
+
+			*msg.buf++ = v & I2C_FIFO_RDATA;
+		}
+		msg.len -= to_read;
+
+		i2c_esp32_reset_fifo(config);
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_write_msg(struct device *dev, u16_t addr,
+			       struct i2c_msg msg)
```

Comment:
There's a possibility it would, yes.  However, both `read` and `write` functions modifies the `msg` parameter (by advancing the `buf` pointer and modifying the `len` variable).  So the copy is safer.